pkg/hooks/hooks.go

```diff
@@ -0,0 +1,70 @@
+package hooks
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/terraform-controller/pkg/provider"
+	"github.com/rancher/terraform-controller/pkg/provider/github"
+	"github.com/rancher/terraform-controller/pkg/types"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+type WebhookHandler struct {
+	providers []provider.Provider
+}
+
+func newHandler(rContext *types.Context) *WebhookHandler {
+	wh := &WebhookHandler{
+		providers: []provider.Provider{
+			// register all supported webhook handler here
+			github.NewGitHub(rContext),
+		},
+	}
+
+	return wh
+}
+
+func (h *WebhookHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	code, err := h.execute(req)
+	if err != nil {
+		e := map[string]interface{}{
+			"type":    "error",
+			"code":    code,
+			"message": err.Error(),
+		}
+		logrus.Debugf("executing webhook request got error: %v", err)
+		rw.WriteHeader(code)
+		responseBody, err := json.Marshal(e)
+		if err != nil {
+			logrus.Errorf("Failed to unmarshall response, error: %v", err)
+		}
+		_, err = rw.Write(responseBody)
+		if err != nil {
+			logrus.Errorf("Failed to write response, error: %v", err)
+		}
+	}
```

Comment:
handler doesn't return 200, so body and code would be sent at the end for function rather than error handling ? 

```
                rw.WriteHeader(code)
		_, err = rw.Write(responseBody)
		if err != nil {
			logrus.Errorf("Failed to write response, error: %v", err)
		}
```